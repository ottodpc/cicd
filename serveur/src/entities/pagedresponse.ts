import { PageInfo } from './pageinfo'

export class PagedResponse {
	results: any[]

	pageInfo: PageInfo

	constructor(results: any[], pageInfo: PageInfo) {
		this.results = results
		this.pageInfo = pageInfo
	}
}
