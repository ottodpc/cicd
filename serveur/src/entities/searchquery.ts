export class SearchQuery {
	filterName: string

	operation: string

	filterValue: any

	constructor(filterName: string, operation: string, filterValue: any) {
		this.filterName = filterName
		this.operation = operation
		this.filterValue = filterValue
	}
}
