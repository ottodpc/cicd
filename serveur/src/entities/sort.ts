export class Sort {
	columnName: string

	orderType: string

	constructor(columnName: string, orderType: string) {
		this.columnName = columnName
		this.orderType = orderType
	}
}
