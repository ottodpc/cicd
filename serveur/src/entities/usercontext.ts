export class UserContext {
	userId: string

	constructor(userId: string) {
		this.userId = userId
	}
}
