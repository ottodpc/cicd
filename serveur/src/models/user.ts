import { Schema } from '@nestjs/mongoose';

import { Prop } from '@nestjs/mongoose';
import { ApiProperty } from '@nestjs/swagger';
import { Todo } from './todo';
import { ValidateNested } from 'class-validator';
import { Type } from 'class-transformer';
import { SchemaFactory } from '@nestjs/mongoose';

@Schema()
export class User {
  @Prop()
  @ApiProperty()
  username: string;

  @Prop()
  @ApiProperty({ type: Todo, isArray: true })
  @ValidateNested({ each: true })
  @Type(() => Todo)
  todos: Todo[];

  @Prop()
  @ApiProperty()
  password: string[];

  @Prop()
  @ApiProperty()
  userId: string;

  @Prop()
  @ApiProperty()
  _id: string;
}

export const UserSchema: SchemaFactory = SchemaFactory.createForClass(User);
