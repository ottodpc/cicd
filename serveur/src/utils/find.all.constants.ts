export const regex = /(\w+):(\w+):(.*)/;

export const allowedModelSorts: Map<string, Set<string>> = new Map();

export const allowedModelFilters: Map<
  string,
  Map<string, string[]>
> = new Map();

export const fieldTypeMapper: Map<string, Map<string, string>> = new Map();

export const Operations = {
  eq: '$eq',
  gt: '$gt',
  gte: '$gte',
  lt: '$lt',
  lte: '$lte',
};

const todoModelProperty: Map<string, string> = new Map();

const allowedTodoFilters: Map<string, string[]> = new Map();

const userModelProperty: Map<string, string> = new Map();

const allowedUserFilters: Map<string, string[]> = new Map();

allowedModelSorts.set('Todo', new Set(['']));
allowedModelSorts.set('User', new Set(['']));
allowedTodoFilters.set('todoId', ['eq']);
allowedModelFilters.set('Todo', allowedTodoFilters);
allowedModelFilters.set('User', allowedUserFilters);
todoModelProperty.set('title', 'string');
todoModelProperty.set('description', 'string');
todoModelProperty.set('todoId', 'string');
fieldTypeMapper.set('Todo', todoModelProperty);
userModelProperty.set('username', 'string');
userModelProperty.set('userId', 'string');
fieldTypeMapper.set('User', userModelProperty);
