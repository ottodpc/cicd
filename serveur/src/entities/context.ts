import { UserContext } from './usercontext'

export class Context {
	userContext: UserContext

	isAuthenticated: boolean

	constructor(userContext: UserContext, isAuthenticated: boolean) {
		this.userContext = userContext
		this.isAuthenticated = isAuthenticated
	}
}
