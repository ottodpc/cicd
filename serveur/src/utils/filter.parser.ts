import { regex } from './find.all.constants'
import { fieldTypeMapper } from './find.all.constants'
import { isFilterOperationValid } from './validator'
import { ParserException } from '../exceptions/parser.exception'
import { SearchQuery } from '../entities/searchquery'

export const filterParser = (
	filters: string,
	modelName: string
): SearchQuery[] => {
	try {
		const searchQuery: SearchQuery[] = []
		const errorMessages: string[] = []
		if (filters === '') {
			return searchQuery
		}
		filters.split(':and:').forEach((filter) => {
			const matchResult = filter.match(regex)
			const filterName = matchResult[1]
			const operation = matchResult[2]
			const filterValue = getFieldValue(
				filterName,
				matchResult[3],
				modelName
			)
			if (isFilterOperationValid(filterName, operation, modelName)) {
				searchQuery.push(
					new SearchQuery(filterName, operation, filterValue)
				)
			} else {
				errorMessages.push(
					`operation ${operation} is not allowed on ${filterName}`
				)
			}
		})

		if (errorMessages.length > 0) {
			throw new ParserException(errorMessages.join(', \n'))
		}
		return searchQuery
	} catch (e) {
		throw new ParserException('failed to parse')
	}
}

const getFieldValue = (
	fieldName: string,
	fieldValue: any,
	modelName: string
) => {
	const modelTypeMapper = fieldTypeMapper.get(modelName) || new Map()
	const type = modelTypeMapper.get(fieldName) || ''

	if (type === 'string') {
		return fieldValue.toString()
	} else if (type === 'number') {
		return parseInt(fieldValue, 10)
	} else if (type === 'Date') {
		return new Date(fieldValue)
	} else if (type === 'Boolean') {
		return fieldValue.toString().toLowerCase() == 'true'
	}
	return fieldValue.toString()
}
