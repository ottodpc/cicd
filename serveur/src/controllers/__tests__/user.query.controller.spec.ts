import { UserService } from '../../services/user.service';
import { UserQueryController } from '../user.query.controller';
import { readResourceFromExactPath } from '../../../test/utils';
import { UserContext } from '../../entities/usercontext';
import { Context } from '../../entities/context';
describe('should test userQueryController', () => {
  let userQueryController: UserQueryController;
  let userService: UserService;

  beforeEach(async () => {
    userService = {
      get: jest.fn(),
      select: jest.fn(),
    } as unknown as UserService;
    userQueryController = new UserQueryController(userService);
  });

  it('should test get method', async () => {
    jest
      .spyOn(userService, 'get')
      .mockReturnValue(
        Promise.resolve(
          await JSON.parse(
            readResourceFromExactPath(__dirname + '/resources/User.json'),
          ),
        ),
      );
    const actualResp = await userQueryController.get(
      'defaultString',
      new Context(new UserContext('defaultString'), false),
    );
    expect(jest.spyOn(userService, 'get')).toBeCalledTimes(1);
    expect(actualResp).toMatchSnapshot();
  });

  it('should test select method', async () => {
    jest
      .spyOn(userService, 'select')
      .mockReturnValue(
        Promise.resolve(
          await JSON.parse(
            readResourceFromExactPath(__dirname + '/resources/User.json'),
          ),
        ),
      );
    const actualResp = await userQueryController.select(
      'defaultString',
      'defaultString',
      123,
      123,
      new Context(new UserContext('defaultString'), false),
    );
    expect(jest.spyOn(userService, 'select')).toBeCalledTimes(1);
    expect(actualResp).toMatchSnapshot();
  });
});
