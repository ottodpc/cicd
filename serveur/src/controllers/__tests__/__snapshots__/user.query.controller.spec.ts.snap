// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test userQueryController should test get method 1`] = `
{
  "_id": "fugiat Lorem irure",
  "fugiat00": -16000000,
  "password": [
    "@Bsjkjs67614562hhs",
  ],
  "todos": [
    {
      "description": "fugiat Lorem irure",
      "fugiat00": -16000000,
      "title": "fugiat Lorem irure",
      "todoId": "fugiat Lorem irure",
    },
  ],
  "userId": "fugiat Lorem irure",
  "username": "fugiat Lorem irure",
}
`;

exports[`should test userQueryController should test select method 1`] = `
PagedResponse {
  "pageInfo": PageInfo {
    "limit": 123,
    "offset": 123,
  },
  "results": {
    "_id": "fugiat Lorem irure",
    "fugiat00": -16000000,
    "password": [
      "@Bsjkjs67614562hhs",
    ],
    "todos": [
      {
        "description": "fugiat Lorem irure",
        "fugiat00": -16000000,
        "title": "fugiat Lorem irure",
        "todoId": "fugiat Lorem irure",
      },
    ],
    "userId": "fugiat Lorem irure",
    "username": "fugiat Lorem irure",
  },
}
`;
